File: Spigot/src/main/java/com/fourteener/megaserver/FileIO.java
package com.fourteener.megaserver;

import java.io.BufferedInputStream;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.FilenameFilter;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.file.FileVisitResult;
import java.nio.file.Files;
import java.nio.file.NoSuchFileException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.SimpleFileVisitor;
import java.nio.file.StandardCopyOption;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.Arrays;
import java.util.List;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

public class FileIO {
	
	// Move a file or folder
	public static void copyFile(String src, String dest, boolean folder) throws IOException {
		if (!folder) {
			Files.copy(Paths.get(src), Paths.get(dest), StandardCopyOption.REPLACE_EXISTING);
		}
		else {
			copyDirectory(src, dest);
		}
	}
	
	// Delete a file or folder
	public static void deleteFile(String src, boolean folder) throws IOException {
		if (!folder) {
			Files.delete(Paths.get(src));
		}
		else {
			delDirectory(src);
		}
	}
	
	// Makes a path of folders
	public static void makePath(String path) {
		File f = new File(path);
		f.mkdirs();
	}
	
	// Zip a folder
	public static void zipDirectory (File directoryToCompress, File outputDirectory, String topFolderName, String outFile) throws InterruptedException{
        try {
            FileOutputStream dest = new FileOutputStream(new File(outputDirectory, outFile + ".zip"));
            ZipOutputStream zipOutputStream = new ZipOutputStream(dest);
            zipDirectoryHelper(directoryToCompress, directoryToCompress, zipOutputStream, topFolderName);
            zipOutputStream.close();
        } catch (Exception e) {
        	System.out.println("Zip error");
        	Thread.sleep(2000);
        }
    }
	
	// List all files at path
	public static List<String> listFiles(String path, boolean folders) {
		if (!folders) {
			File f = new File(path);
			return Arrays.asList(f.list());
		}
		File file = new File(path);
		String[] directories = file.list(new FilenameFilter() {
			@Override
			public boolean accept(File current, String name) {
				return new File(current, name).isDirectory();
			}
		});
		return Arrays.asList(directories);
	}
	
	// Read a text file into a string
	public static String readFromFile(String file) throws IOException {
		try {
			List<String> fileContentsList = Files.readAllLines(Paths.get(file));
			String fileContents = "";
			for (String i : fileContentsList) {
				fileContents += i + "\n";
			}
			return fileContents;
		}
		catch (NoSuchFileException e) {
			return "";
		}
	}
	
	// File exists?
	public static boolean exists(String path) {
		File f = new File(path);
		return f.exists();
	}
	
	// Writes a string to a file
	public static void writeToFile (String file, boolean append, String text) throws IOException {
		FileWriter fileWriter = new FileWriter (file, append);
		BufferedWriter bufferedWriter = new BufferedWriter (fileWriter);
		bufferedWriter.write(text);
		bufferedWriter.close();
		fileWriter.close();
	}
	
	// Writes the data in textList to file, splitting each element with sepChar
	public static void writeToFile (String file, boolean append, List<String> textList, String sepChar) throws IOException {
		String text = "";
		for (String s : textList) {
			text += s + sepChar;
		}
		FileWriter fileWriter = new FileWriter (file, append);
		BufferedWriter bufferedWriter = new BufferedWriter (fileWriter);
		bufferedWriter.write(text);
		bufferedWriter.close();
		fileWriter.close();
	}
	
	// Used to delete a folder with files
	private static void delDirectory (String target) throws IOException {
		Path directory = Paths.get(target);
		Files.walkFileTree(directory, new SimpleFileVisitor<Path>() {
			@Override
			public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
				Files.delete(file);
				return FileVisitResult.CONTINUE;
			}
			@Override
			public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
				Files.delete(dir);
				return FileVisitResult.CONTINUE;
			}
		});
	}
	
	// Helper for zipping folders
    private static void zipDirectoryHelper (File rootDirectory, File currentDirectory, ZipOutputStream out, String topFolderName) throws Exception {
        byte[] data = new byte[2048];
        File[] files = currentDirectory.listFiles();
        if (files == null) {
          // no files were found or this is not a directory
        } else {
            for (File file : files) {
                if (file.isDirectory() && !file.getName().equals("")) {
                    zipDirectoryHelper(rootDirectory, file, out, topFolderName);
                } else {
                    FileInputStream fi = new FileInputStream(file);
                    // creating structure and avoiding duplicate file names
                    String name = file.getAbsolutePath().replace(rootDirectory.getAbsolutePath(), topFolderName);

                    ZipEntry entry = new ZipEntry(name);
                    out.putNextEntry(entry);
                    int count;
                    BufferedInputStream origin = new BufferedInputStream(fi,2048);
                    while ((count = origin.read(data, 0 , 2048)) != -1){
                        out.write(data, 0, count);
                    }
                    origin.close();
                }
            }
        }
    }
    
    // Used to copy a directory (replacing existing files)
 	private static void copyDirectory (String src, String dest) throws IOException {
 		File source = new File (src);
 		File destination = new File (dest);
 	    if (source.isDirectory()) {
 	        if (!destination.exists()) {
 	            destination.mkdirs();
 	        }
 	        String files[] = source.list();
 	        for (String file : files) {
 	            File srcFile = new File(source, file);
 	            File destFile = new File(destination, file);
 	            copyDirectory(srcFile.toString(), destFile.toString());
 	        }
 	    } else {
 	        InputStream in = null;
 	        OutputStream out = null;
 	        try {
 	            in = new FileInputStream(source);
 	            out = new FileOutputStream(destination);
 	            byte[] buffer = new byte[1024];
 	            int length;
 	            while ((length = in.read(buffer)) > 0)
 	            {
 	                out.write(buffer, 0, length);
 	            }
 	        } catch (Exception e) {
 	            // Just ignore the error, it's to be expected
 	        }
 	        in.close();
 	        out.close();
 	    }
 	}
}

File: src/main/java/com/_14ercooper/megaserver/FileIO.kt
/**
 * This file is part of 14erEdit.
 *
  * 14erEdit is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * 14erEdit is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with 14erEdit.  If not, see <https://www.gnu.org/licenses/>.
 */

package com._14ercooper.megaserver

import java.io.*
import java.nio.file.*
import java.nio.file.attribute.BasicFileAttributes
import java.util.zip.CRC32
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream

object FileIO {
    // Move a file or folder
    @Throws(IOException::class)
    fun copyFile(src: String, dest: String, folder: Boolean) {
        if (!folder) {
            Files.copy(Paths.get(src), Paths.get(dest), StandardCopyOption.REPLACE_EXISTING)
        } else {
            copyDirectory(src, dest)
        }
    }

    // Delete a file or folder
    @Throws(IOException::class)
    fun deleteFile(src: String, folder: Boolean) {
        if (!folder) {
            Files.deleteIfExists(Paths.get(src))
        } else {
            delDirectory(src)
        }
    }

    // Makes a path of folders
    fun makePath(path: String) {
        File(path).mkdirs()
    }

    // Zip a folder
    @Throws(InterruptedException::class)
    fun zipDirectory(
        directoryToCompress: File, outputDirectory: File, topFolderName: String,
        outFile: String
    ) {
        try {
            val dest = FileOutputStream(File(outputDirectory, "$outFile.zip"))
            val zipOutputStream = ZipOutputStream(dest)
            zipDirectoryHelper(directoryToCompress, directoryToCompress, zipOutputStream, topFolderName)
            zipOutputStream.close()
        } catch (e: Exception) {
            println("Zip error")
            Thread.sleep(2000)
        }
    }

    // List all files at path
    fun listFiles(path: String, folders: Boolean): List<String> {
        if (!folders) {
            val f = File(path)
            return listOf(*f.list()!!)
        }
        val file = File(path)
        val directories = file.list { current, name -> File(current, name).isDirectory }
        return listOf(*directories!!)
    }

    // Read a text file into a string
    @Throws(IOException::class)
    fun readFromFile(file: String): String {
        return try {
//            val fileContentsList = Files.readAllLines(Paths.get(file))
            Files.readString(Paths.get(file))
//            for (i in fileContentsList) {
//                fileContents += """
//                $i
//
//                """.trimIndent()
//            }
//            fileContents
        } catch (e: NoSuchFileException) {
            ""
        }
    }

    // File exists?
    fun exists(path: String): Boolean {
        return File(path).exists()
    }

    // Writes a string to a file
    @Throws(IOException::class)
    fun writeToFile(file: String, append: Boolean, text: String) {
        val fileWriter = FileWriter(file, append)
        val bufferedWriter = BufferedWriter(fileWriter)
        bufferedWriter.write(text)
        bufferedWriter.close()
        fileWriter.close()
    }

    // Used to delete a folder with files
    @Throws(IOException::class)
    private fun delDirectory(target: String) {
        val directory = Paths.get(target)
        Files.walkFileTree(directory, object : SimpleFileVisitor<Path>() {
            @Throws(IOException::class)
            override fun visitFile(file: Path, attrs: BasicFileAttributes): FileVisitResult {
                Files.delete(file)
                return FileVisitResult.CONTINUE
            }

            @Throws(IOException::class)
            override fun postVisitDirectory(dir: Path, exc: IOException?): FileVisitResult {
                Files.delete(dir)
                return FileVisitResult.CONTINUE
            }
        })
    }

    // Helper for zipping folders
    @Throws(Exception::class)
    private fun zipDirectoryHelper(
        rootDirectory: File, currentDirectory: File, out: ZipOutputStream,
        topFolderName: String
    ) {
        val data = ByteArray(2048)
        val files = currentDirectory.listFiles()
        if (files == null) {
            // no files were found or this is not a directory
        } else {
            for (file in files) {
                if (file.isDirectory && file.name != "") {
                    zipDirectoryHelper(rootDirectory, file, out, topFolderName)
                } else {
                    val fi = FileInputStream(file)
                    // creating structure and avoiding duplicate file names
                    val name = file.absolutePath.replace(rootDirectory.absolutePath, topFolderName)
                    val entry = ZipEntry(name)
                    out.putNextEntry(entry)
                    var count: Int
                    val origin = BufferedInputStream(fi, 2048)
                    while (origin.read(data, 0, 2048).also { count = it } != -1) {
                        out.write(data, 0, count)
                    }
                    origin.close()
                }
            }
        }
    }

    // Used to copy a directory (replacing existing files)
    @Throws(IOException::class)
    private fun copyDirectory(src: String, dest: String) {
        val source = File(src)
        val destination = File(dest)
        if (source.isDirectory) {
            if (!destination.exists()) {
                destination.mkdirs()
            }
            val files = source.list()!!
            for (file in files) {
                val srcFile = File(source, file)
                val destFile = File(destination, file)
                copyDirectory(srcFile.toString(), destFile.toString())
            }
        } else {
            var `in`: InputStream? = null
            var out: OutputStream? = null
            try {
                `in` = FileInputStream(source)
                out = FileOutputStream(destination)
                val buffer = ByteArray(1024)
                var length: Int
                while (`in`.read(buffer).also { length = it } > 0) {
                    out.write(buffer, 0, length)
                }
            } catch (e: Exception) {
                // Just ignore the error, it's to be expected
            }
            `in`!!.close()
            out!!.close()
        }
    }

    fun getHash(file: String): String {
        Files.newInputStream(Paths.get(file)).use {
            val crc = CRC32()
            var c = 0
            while (c != -1) {
                c = it.read()
                crc.update(c)
            }
            return Integer.toHexString(crc.value.toInt())
        }
    }
}